source/FreeRTOS_IP.c

```diff
@@ -1544,6 +1544,9 @@ static void prvProcessEthernetPacket( NetworkBufferDescriptor_t * const pxNetwor
     eFrameProcessingResult_t eReturned = eReleaseBuffer;
 
     configASSERT( pxNetworkBuffer != NULL );
+    configASSERT( pxNetworkBuffer->pxInterface != NULL );
+    configASSERT( pxNetworkBuffer->pxEndPoint != NULL );
```

Comment:
>     configASSERT( pxNetworkBuffer != NULL );
>     configASSERT( pxNetworkBuffer->pxInterface != NULL );

Some people may wonder why a network buffer has both a reference to an `pxEndPoint` and to an `pxInterface`, because when you know the `pxEndPoint`, `pxInterface` is also known.
The answer is that the reverse is not true: when a packet is received on interface `pxInterface`, we still have to find out which endpoint is handling the incoming packets.

About the added `configASSERT`'s: yes I agree, it is in line with the idea of keeping the initial `pxEndPoint` that was obtained by the network interface.

Reply:
Isn't this a way that @evpopov post [here](https://forums.freertos.org/t/freertos-tcp-routing-from-interface-to-interface/19112/5) would be implemented? Something like `pxNetworkBuffer->pxInterface` and `pxNetworkBuffer->pxInterface->pxEndPoint` are the interfaces and endpoints received from, and `pxNetworkBuffer->pxEndPoint` and `pxNetworkBuffer->pxEndPoint->pxNetworkInterface` are the target interfaces and endpoints?